c10/core/impl/cow/try_ensure.cpp

```diff
@@ -0,0 +1,102 @@
+#include <c10/core/impl/cow/try_ensure.h>
+
+#include <c10/core/Allocator.h>
+#include <c10/core/StorageImpl.h>
+#include <c10/core/impl/cow/context.h>
+#include <c10/core/impl/cow/deleter.h>
+#include <c10/util/Exception.h>
+#include <c10/util/UniqueVoidPtr.h>
+
+#include <memory>
+#include <optional>
+
+namespace c10::impl {
+
+namespace {
+
+// A std::unique_ptr deleter that is expected to never execute.
+struct AssertFalseDeleter {
+  template <typename T>
+  auto operator()(T* /* ptr */) const -> void {
+    TORCH_INTERNAL_ASSERT(false); // expected to never be called
+  }
+};
+
+// Wraps a DataPtr with a copy-on-write DataPtr.
+auto make_data_ptr(at::DataPtr const& data_ptr, cow::Context& ctx)
+    -> at::DataPtr {
+  return at::DataPtr(
+      data_ptr.get(), &ctx, cow::delete_context, data_ptr.device());
```

Comment:
I was expecting to see some reference to the original data pointers deleter here?

Reply:
Looks like it's handled below